enhancements/kube-apiserver/stale-clusteroperators-detection.md

```diff
@@ -0,0 +1,117 @@
+---
+title: detection-and-reporting-of-stale-clusteroperators
+authors:
+  - "@mfojtik"
+reviewers:
+  - "@deads2k"
+approvers:
+  - "@tkashem"
+creation-date: 2022-01-31
+last-updated: 2023-02-13
+---
+
+# Detection and reporting of stale cluster operators
+
+## Summary
+
+In OpenShift 4, every active operator included in the core payload (cluster operator) updates its conditions in "
+clusteroperator" resource. Conditions determine whether specific core components are Degraded or Available or if they
+are making progress or if their state causes the cluster to be not upgradeable. These conditions are essential for
+cluster administrators to determine the health of their clusters, and they are usually the first thing Red Hat support
+is going to look at in case there is a problem with the cluster.
+
+Unfortunately, in some edge cases, these conditions might not reflect the current state of the cluster. For example,
+since they depend on the running operator, these conditions might become stale and outdated if the operator is not
+running or cannot contact the API server. And while the API server might still work, the result
+of `oc get clusteroperators` does not represent the cluster's current state.
+
+Therefore, we need a detection mechanism to help detect in-active operators or operators that are not actively updating
+the conditions.
+
+## Motivation
+
+In the past, there were multiple cases when either kubelet or CRI-O failed to run pods, either by KCM malfunctioning or
+some other system error. However, the static pods succeeded in running, making the Kubernetes API server available and
+accessible with `oc`. When support or engineers ran `oc get clusteroperators` against this cluster, they saw all
+operators reporting available, non-degraded and upgradeable status, but the truth was that none of the operators that
+provided these conditions were running.
```

Comment:
I am interested to loo in to the specific bug or bugs. My intent is to see why the operator reconciliation logic failed to bring out the issues.

Reply:
In at least a few of the cases I looked at the operators were not running properly after hibernation. But it could be anything that leads to an operator crashlooping before it's able to accurately update CO Conditions.